plugins/jabra-gnp/fu-jabra-gnp-device.c

```diff
@@ -0,0 +1,859 @@
+					     NULL,
+					     receive_data->timeout,
+					     NULL, /* cancellable */
+					     error)) {
+		g_prefix_error(error, "failed to read from device: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_gnp_device_receive_with_sequence_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuJabraGnpDevice *self = FU_JABRA_GNP_DEVICE(device);
+	FuJabraGnpReceiveData *receive_data = (FuJabraGnpReceiveData *)user_data;
+
+	if (!fu_jabra_gnp_device_receive_cb(device, user_data, error))
+		return FALSE;
+	if (self->sequence_number != receive_data->rxbuf[3]) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "sequence_number error -- got 0x%x, expected 0x%x",
+			    receive_data->rxbuf[3],
+			    self->sequence_number);
+		return FALSE;
+	}
+	self->sequence_number += 1;
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_gnp_device_read_dfu_pid(FuJabraGnpDevice *self, GError **error)
+{
+	guint8 txbuf[FU_JABRA_GNP_BUF_SIZE] =
+	    {FU_JABRA_GNP_IFACE, 0x08, 0x00, self->sequence_number, 0x46, 0x02, 0x13};
+	guint8 rxbuf[FU_JABRA_GNP_BUF_SIZE] = {0x00};
+	FuJabraGnpSendData send_data = {.txbuf = txbuf,
+					.timeout = FU_JABRA_GNP_STANDARD_SEND_TIMEOUT};
+	FuJabraGnpReceiveData receive_data = {.rxbuf = rxbuf,
+					      .timeout = FU_JABRA_GNP_STANDARD_RECEIVE_TIMEOUT};
+
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_gnp_device_send_cb,
+				  FU_JABRA_GNP_MAX_RETRIES,
+				  FU_JABRA_GNP_RETRY_DELAY,
+				  (gpointer)&send_data,
+				  error))
+		return FALSE;
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_gnp_device_receive_with_sequence_cb,
+				  FU_JABRA_GNP_MAX_RETRIES,
+				  FU_JABRA_GNP_RETRY_DELAY,
+				  (gpointer)&receive_data,
+				  error))
+		return FALSE;
+	self->dfu_pid = fu_memread_uint16(receive_data.rxbuf + 7, G_LITTLE_ENDIAN);
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_gnp_device_read_version(FuJabraGnpDevice *self, GError **error)
+{
+	guint8 txbuf[FU_JABRA_GNP_BUF_SIZE] =
+	    {FU_JABRA_GNP_IFACE, 0x08, 0x00, self->sequence_number, 0x46, 0x02, 0x03};
+	guint8 rxbuf[FU_JABRA_GNP_BUF_SIZE] = {0x00};
+	FuJabraGnpSendData send_data = {.txbuf = txbuf,
+					.timeout = FU_JABRA_GNP_STANDARD_SEND_TIMEOUT};
+	FuJabraGnpReceiveData receive_data = {.rxbuf = rxbuf,
+					      .timeout = FU_JABRA_GNP_STANDARD_RECEIVE_TIMEOUT};
+	g_autofree gchar *version = NULL;
+
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_gnp_device_send_cb,
+				  FU_JABRA_GNP_MAX_RETRIES,
+				  FU_JABRA_GNP_RETRY_DELAY,
+				  (gpointer)&send_data,
+				  error))
+		return FALSE;
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_gnp_device_receive_with_sequence_cb,
+				  FU_JABRA_GNP_MAX_RETRIES,
+				  FU_JABRA_GNP_RETRY_DELAY,
+				  (gpointer)&receive_data,
+				  error))
+		return FALSE;
+	version = fu_strsafe((const gchar *)receive_data.rxbuf + 8, FU_JABRA_GNP_BUF_SIZE - 8);
+	fu_device_set_version(FU_DEVICE(self), version);
```

Comment:
Is this always going to be a "triplet"? e.g. X.Y.Z? If you use PLAIN verfmt that's fine, but it gets nasty when you compare `1.12.0` and `1.2.0` as ASCII strings.

Reply:
I wasn't entirely sure if `1.12.0` would classify was a triplet! As we can have firmware in which the minor or micro can go above 9. If that's okay in the triplet format, it should cover our version formatting.

Reply:
yup, triplet can be XXXXX.YYYYYY.ZZZZZZ i.e. pretty big -- and it'll massively simplify version comparisons. Thanks.